### ISSUE TITLE
How to setup Nested Form Groups in the APM updated?

### ISSUE THREAD

**Issue by dmarg** [DeborahK/Angular2-ReactiveForms#5]
Hi, 

I watched your course on pluralsight which was awesome. Thanks for doing that one. I am confused on how to set up nested form groups to then have the generic validator process the messages. 

You have this: 

this.validationMessages = {
            productName: {
                required: 'Product name is required.',
                minlength: 'Product name must be at least three characters.',
                maxlength: 'Product name cannot exceed 50 characters.'
            },
            productCode: {
                required: 'Product code is required.'
            },
            starRating: {
                range: 'Rate the product between 1 (lowest) and 5 (highest).'
            }
        };

        // Define an instance of the validator for use with this form, 
        // passing in this form's set of validation messages.
        this.genericValidator = new GenericValidator(this.validationMessages);

// product form 
 this.productForm = this.fb.group({
            productName: ['', [Validators.required,
                               Validators.minLength(3),
                               Validators.maxLength(50)]],
            productCode: ['', Validators.required],
            starRating: ['', NumberValidators.range(1, 5)],
            tags: this.fb.array([]),
            description: ''
        });

I tried doing something similar but we have nested form groups but I can't get it to work. 

for example if the this.productForm looked like this with productCode being nested:
 this.productForm = this.fb.group({
            productName: ['', [Validators.required,
                               Validators.minLength(3),
                               Validators.maxLength(50)]],
            productCode: this.fb.group({
              internal: ['', Validators.required],              
              external: ['', Validators.required]              
            }),
            starRating: ['', NumberValidators.range(1, 5)],
            tags: this.fb.array([]),
            description: ''
        });

How would you write the this.validationMessages? I tried a nested object like this but it did not work.
this.validationMessages = {
            productName: {
                required: 'Product name is required.',
                minlength: 'Product name must be at least three characters.',
                maxlength: 'Product name cannot exceed 50 characters.'
            },
            productCode: {
                internal: {
                  required: 'Internal Product code is required.'
                }
                external: {
                  required: 'External Product code is required.'
                }
            },
            starRating: {
                range: 'Rate the product between 1 (lowest) and 5 (highest).'
            }
        };

Any suggestions?

**Comment by dmarg** [comment]
figured out a work around but don't know how to process message for the extra validators.

**Comment by DeborahK** [comment]
You are correct ... I did not add code to support generic validation of nested form groups. At the time I did the course, I did not have a solution.

I'd love to see what you have come up with if you have your code posted to github somewhere ...

**Comment by dmarg** [comment]
So, I realized that the generic validator for the message processor was recursive if it had child form groups. I needed something for a confirm password field and on the child group I couldn't put an extra (sync) validator for password match and get it to process the messages.

So what I did was move the password and confirm password to the main form group and then after setup I set a new validator .

On the component this how I set up the form and then set the extra validator on confirmPassword after the form was set up.

``` JavaScript
    this.registerPasswordForm = this.formBuilder.group({
      password: ['', [Validators.required, Validators.pattern('^(?=.*[A-Z])(?=.*[0-9])(?=.*[a-z]).{8,30}$')]],
      confirmPassword: [''],
    });

    this.registerPasswordForm.get('confirmPassword').setValidators([Validators.required, PasswordValidators.passwordMatch]);

    this.validationMessages = {
      password: {
        required: 'required',
        pattern: 'invalid password'
      },
      confirmPassword: {
        required: 'required',
        passwordMatch: 'passwords do not match'
      }
    };
    this.messageProcessor = new MessageProcessor(this.validationMessages);
```

Then the PasswordValidator file looked like this:
``` JavaScript
import { AbstractControl } from '@angular/forms';

export class PasswordValidators {

  static passwordMatch(control: AbstractControl): { [key: string]: boolean } {
    if (control.parent.get('password').value !== control.value) {
      return {
        'passwordMatch': true
      };
    }
    return null;
  }

}
```

I don't know if this is the best way but it works. What are your thoughts?



Lastly, I did it the way above because I was not entirely sure how to get the validator message of passwordMatch to display correctly on the view if the password form group looked like this.
```JavaScript
this.registerPasswordForm = this.formBuilder.group({
      password: ['', [Validators.required, Validators.pattern('^(?=.*[A-Z])(?=.*[0-9])(?=.*[a-z]).{8,30}$')]],
      confirmPassword: ['', [Validators.required]],
    }, {validator: passwordMatch});
```

**Comment by dmarg** [comment]
@DeborahK Also, thanks so much for responding!

**Comment by DeborahK** [comment]
When I was trying this out a few months ago ... one problem I ran into is that the validation error is on the group, not the control. So that makes it more difficult for a generic validator to add the message to the appropriate element.

I came to the conclusion that form groups may need to be handled manually, outside of any generic validator code, in order for it to work correctly.

It looks like you may have gotten past that?

**Comment by dmarg** [comment]
I had issues with the group validator as well. I really liked your approach with the generic validator message processor. And I tried a bunch of different things to get the validation error from the group to be processed through that and it did not work.

I did what I did above where I set the form group and then setValidators on a form control with in that group after the group was set up because then I could access the parent of that form control which is the form group and then go access the other attribute I needed to validate against.

I don't think it is "getting past" the form group validator but just using the parent on the form control to then access the form group and create a custom validator.

Again, thanks for responding. Also, your course on reactive forms on pluralsight was very helpful and instrumental for our team. Thanks!
